File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
import "math"

func diameterOfBinaryTree(root *TreeNode) int {
    var diameterOfBT float64;
    
    if (root == nil) {
        return 0;
    }
    dfsCountEdges(root, &diameterOfBT)
    return int (diameterOfBT);
}


func dfsCountEdges(n *TreeNode, diameterOfBT *float64) float64 {
    
    var left float64;
    var right float64;
    if(n == nil) {
        return 0
    }
    left = dfsCountEdges(n.Left, diameterOfBT)
    right = dfsCountEdges(n.Right, diameterOfBT)
    *diameterOfBT = math.Max(*diameterOfBT, left+right)
    return math.Max(left, right) + 1
    
}

File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.py
"""
1- Need to get the Hight and Diametter of the binary tree
2- Need to use DFS Algo to compute the Hight and also Diametter
3- Need to use recursive Datastructure
4- Time complexity: O(n), Space: O(n)

"""

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        diameter = [0]
        
        def dfs(root):
                if root is None:
                    return -1
            
                left = dfs(root.left)
                right = dfs(root.right)
                diameter[0] = max(diameter[0], 2 + left + right) #Computing B.T diammeter
                return (1 + max(left, right)) # returning hight
        
        dfs(root)
        return diameter[0]
             
            
        